Fix various DB Query issues

jvehent opened the pull request:
@ameihm0912 : r?

ameihm0912 commented:
It looks good to me, except it looks like the defer call for rows.Close() should happen after we know we have a valid rows or statement value.

jvehent commented:
I fixed the `defer rows.Close()` to only happen if `rows != nil`.
I also added a fairly beefy patch to build smarter search queries, in the hope that it will improve search perfs.
